### PR TITLE
Refactor settings in errors and user_id_generator (Part 4)

### PR DESCRIPTION
**Pull request by samiwel**
### What is the context of this PR?
This PR is a continuation of the work to refactor survey runner to use settings from Flask rather than referencing the settings module directly.
In particular, this PR is concerned with refactoring the following modules:
- `errors.py`
- `user_id_generator.py`

### How to review 
All existing automated tests should continue to run and pass. External behaviour should remain unchanged.

### CODE REVIEW

**Review by adoublebarrel — APPROVED** [review]
Looks good to me.

File: app/authentication/user_id_generator.py [hunk 1]
```diff
@@ -15,28 +14,28 @@
 
 class UserIDGenerator(object):
 
-    @staticmethod
-    def generate_id(metadata):
+    def __init__(self, iterations):
+        self.iterations = iterations
```
Comment (collisdigital) [inline app/authentication/user_id_generator.py]: make this a private var?

File: app/authentication/user_id_generator.py [hunk 2]
```diff
@@ -15,28 +14,28 @@
...
+
+    def generate_id(self, metadata, salt):
         collection_exercise_sid, eq_id, form_type, ru_ref = UserIDGenerator._get_token_data(metadata)
```
Comment (collisdigital) [inline app/authentication/user_id_generator.py]: Worth adding a None check for metadata and salt and throwing `ValueError` (plus unit test) as this is critical stuff

File: app/authentication/user_id_generator.py [hunk 3]
```diff
@@ -15,28 +14,28 @@
...
         logger.debug("generating user id", ru_ref=ru_ref, ce_id=collection_exercise_sid, eq_id=eq_id, form_type=form_type)
-        salt = to_bytes(settings.EQ_SERVER_SIDE_STORAGE_USER_ID_SALT)
-        user_id = UserIDGenerator._generate(collection_exercise_sid, eq_id, form_type, ru_ref, salt)
+        salt = to_bytes(salt)
+        user_id = self._generate(collection_exercise_sid, eq_id, form_type, ru_ref, salt)
         return to_str(user_id)
 
-    @staticmethod
-    def generate_ik(token):
+    def generate_ik(self, token, salt):
```
Comment (collisdigital) [inline app/authentication/user_id_generator.py]: Check inputs as above

**Review by collisdigital — CHANGES_REQUESTED** [review]

File: app/authentication/user_id_generator.py (hunk 3 above)
Comment (samiwel) [reply app/authentication/user_id_generator.py]: Inputs are now checked in my latest commit.

File: app/authentication/authenticator.py [hunk 4]
```diff
@@ -61,8 +61,9 @@ def store_session(metadata):
     session.clear()
 
     # get the hashed user id for eq
-    user_id = UserIDGenerator.generate_id(metadata)
-    user_ik = UserIDGenerator.generate_ik(metadata)
+    id_generator = UserIDGenerator(current_app.config['EQ_SERVER_SIDE_STORAGE_USER_ID_ITERATIONS'])
```
Comment (jonnyshaw89) [inline app/authentication/authenticator.py]: I think we should pass in `EQ_SERVER_SIDE_STORAGE_USER_ID_SALT` and `EQ_SERVER_SIDE_STORAGE_USER_IK_SALT` into the constructor as they never change during runtime

File: app/views/flush.py [hunk 5]
```diff
@@ -70,6 +70,7 @@ def _submit_data(user):
 
 
 def _get_user(decrypted_token):
-    user_id = UserIDGenerator.generate_id(decrypted_token)
-    user_ik = UserIDGenerator.generate_ik(decrypted_token)
+    id_generator = UserIDGenerator(current_app.config['EQ_SERVER_SIDE_STORAGE_USER_ID_ITERATIONS'])
```
Comment (jonnyshaw89) [inline app/views/flush.py]: As we are constructing this twice with the same arguments i think we should construct it once on application startup.

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by jonnyshaw89** [comment]
@collisdigital your blocking this one now.

**Review by collisdigital — APPROVED** [review]
Please start a session on master, switch to this branch and check the data is restored correctly.

**Comment by samiwel** [comment]
@collisdigital Chris, just checked this and can confirm session data was restored correctly. I'll squash the commits down now

**Comment by collisdigital** [comment]
👍

**Pull request merged**